executor/new_distsql.go

```diff
@@ -189,10 +201,202 @@ func (e *IndexReaderExecutor) doRequest() error {
 	}
-	e.result, err = distsql.SelectDAG(e.ctx.GetClient(), goctx.Background(), e.dagPB, kvRanges, e.ctx.GetSessionVars().DistSQLScanConcurrency, e.keepOrder, e.desc)
+	e.result, err = distsql.SelectDAG(e.ctx.GetClient(), e.ctx.GoCtx(), e.dagPB, kvRanges, e.ctx.GetSessionVars().DistSQLScanConcurrency, e.keepOrder, e.desc)
 	if err != nil {
 		return errors.Trace(err)
 	}
 	e.result.Fetch(e.ctx.GoCtx())
 	return nil
 }
+
+// IndexLookUpExecutor implements double read for index scan.
+type IndexLookUpExecutor struct {
+	asName    *model.CIStr
+	table     table.Table
+	index     *model.IndexInfo
+	tableID   int64
+	keepOrder bool
+	desc      bool
+	ranges    []*types.IndexRange
+	dagPB     *tipb.DAGRequest
+	ctx       context.Context
+	schema    *expression.Schema
+
+	// result returns one or more distsql.PartialResult.
+	result distsql.SelectResult
+
+	taskChan chan *lookupTableTask
+	tasksErr error
+	taskCurr *lookupTableTask
+
+	tableRequest *tipb.DAGRequest
+}
+
+func (e *IndexLookUpExecutor) doRequest() error {
+	fieldTypes := make([]*types.FieldType, len(e.index.Columns))
+	for i, v := range e.index.Columns {
+		fieldTypes[i] = &(e.table.Cols()[v.Offset].FieldType)
+	}
+	kvRanges, err := indexRangesToKVRanges(e.ctx.GetSessionVars().StmtCtx, e.tableID, e.index.ID, e.ranges, fieldTypes)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.result, err = distsql.SelectDAG(e.ctx.GetClient(), e.ctx.GoCtx(), e.dagPB, kvRanges, e.ctx.GetSessionVars().DistSQLScanConcurrency, e.keepOrder, e.desc)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.result.Fetch(e.ctx.GoCtx())
+
+	// Use a background goroutine to fetch index and put the result in e.taskChan.
+	// e.taskChan serves as a pipeline, so fetching index and getting table data can
+	// run concurrently.
+	e.taskChan = make(chan *lookupTableTask, LookupTableTaskChannelSize)
+	go e.fetchHandles()
+	return nil
+}
+
+// executeTask executes the table look up tasks. We will construct a table reader and send request by handles.
+// Then we hold the returning rows and finish this task.
+func (e *IndexLookUpExecutor) executeTask(task *lookupTableTask) {
+	var err error
+	defer func() {
+		task.doneCh <- err
+	}()
+	tableReader := &TableReaderExecutor{
+		asName:  e.asName,
+		table:   e.table,
+		tableID: e.tableID,
+		dagPB:   e.tableRequest,
+		schema:  e.schema,
+		ctx:     e.ctx,
+	}
+	err = tableReader.doRequestForHandles(task.handles)
+	if err != nil {
+		return
+	}
+	for {
+		var row *Row
+		row, err = tableReader.Next()
+		if err != nil || row == nil {
+			return
+		}
+		task.rows = append(task.rows, row)
+	}
+}
+
+// fetchHandles fetches a batch of handles from index data and builds the index lookup tasks.
+// We initialize some workers to execute this tasks concurrently and put the task to taskCh by order.
+func (e *IndexLookUpExecutor) fetchHandles() {
+	lookupConcurrencyLimit := e.ctx.GetSessionVars().IndexLookupConcurrency
+	workCh := make(chan *lookupTableTask, lookupConcurrencyLimit)
```

Comment:
workCh is a local variable. It's duty is to allocate task to different workers. In old code, It's length is 1. I enlarge this channel to avoid congestion which rarely happens.